### PR TITLE
small changes/fixes for pitch detection algo

### PR DESCRIPTION
**Pull request by paradigmn**
### What does this PR do?
- change the processing order to make the "halftone != -1" condition valid
- removed element wise division by "samplesSinceLastFrame" -> doesn't appear to be necessary for pitch detection

### Additional Notes
It could be worth considering to combine CircularAverageMagnitudeDifference() and CalculateBestFittingHalftone() into one method. This would make the correlation buffer obsolete. Granted the performance gain is not that huge, but in time critical applications moving data unnecessarily should be avoided.

I couldn't test any of the proposed changes. Please check beforehand if anything works as expected if you want to apply them.

### CODE REVIEW

**Comment by achimmihca** [comment]
I just merged another branch into the master. Please, pull and resolve the conflicts.

> removed element wise division by "samplesSinceLastFrame" -> doesn't appear to be necessary for pitch detection

>I couldn't test any of the proposed changes

If you did not test, then how do you know the element wise division is not necessary?

**Comment by basisbit** [comment]
Indeed, currently that devision is not necessary because the output of the camd function is only used to find the pitch that best fits. The devision would only be required if we would use the result to show a spectrum graph for each player, which would be neat, but is not there currently.

File: UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs [hunk 1]
```diff
@@ -85,11 +85,11 @@ public void ProcessAudioSamples(float[] audioSamplesBuffer, int samplesSinceLast
         // get best fitting tone
         double[] correlation = CircularAverageMagnitudeDifference(audioSamplesBuffer, sampleCountToUse);
 
-        // no idea where the +3 is coming from...
-        int halftone = CalculateBestFittingHalftone(correlation) + BaseToneMidi + 3;
+        int halftone = CalculateBestFittingHalftone(correlation);
         if (halftone != -1 && isEnabled)
         {
-            OnPitchDetected(halftone);
+            // add C5 (A4 + 3 halftones) as offset to the detected pitch
```
Comment (basisbit) [inline UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs]: This cide comment is not correct. BaseToneMidi + 3 is being displayed as A4 in the game UI. I just didn't find the code that causes the +3 offset yet. Please adjust this cide comment.

Comment (paradigmn) [reply UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs]: @achimmihca The pitch detection searches for the smallest value in the correlation buffer. Dividing this buffer by a constant doesn't change the minimum.

@basisbit Ah ok, so the reciprocal correlation buffer could then be used as a rough approximation of the signal spectrum. Do you prefer to keep this line for future implementation?

As for the offset: I assumed adding C5 to the detected pitch has the purpose of keeping the midi notation valid. If the pitch detection returns 2 (D0), then adding C5 (36) would result in D5 (38), while an offset of A4 (33) would result in B4 (35). If a detected pitch of 0 (+A4+3) would indeed result in an A4 being displayed at UI level, would imply that a detected 0 is actually an A0 which doesn't align with the algo as far as I can tell.

**Comment by paradigmn** [comment]
@achimmihca The pitch detection searches for the smallest value in the correlation buffer. Dividing this buffer by a constant doesn't change the minimum.

@basisbit Ah ok, so the reciprocal correlation buffer could then be used as a rough approximation of the signal spectrum. Do you prefer to keep this line for future implementation?

As for the offset: I assumed adding C5 to the detected pitch has the purpose of keeping the midi notation valid. If the pitch detection returns 2 (D0), then adding C5 (36) would result in D5 (38), while an offset of A4 (33) would result in B4 (35). If a detected pitch of 0 (+A4+3) would indeed result in an A4 being displayed at UI level, would imply that a detected 0 is actually an A0 which doesn't align with the algo as far as I can tell.

File: UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs (hunk 1 above)
Comment (achimmihca) [reply UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs]: When I look at this comparison of midi note integer, name and frequency: https://newt.phys.unsw.edu.au/jw/notes.html
then I would say BaseToneMidi = 33 and 33+ 3 = 36 and 36 is C2.
So who is right now? Maybe the game UI is not correct. Or BaseToneMidi is wrong (the link above names 69 as concert pitch A4)

Comment (basisbit) [reply UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs]: @achimmihca indeed. I don't understand why when I was coding the camd implementation, it required the +3 to make the Midi code translate it to the correct tone.
If I set halftone = 33 (without the +3), it now shows A1.

so, @paradigmn could you please update your pull request to remove the +3 calculation and remove that  from the comment?

Comment (paradigmn) [reply UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs]: I am not at home right now. I will try to get it done when I am back.

Comment (basisbit) [reply UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs]: @achimmihca please use https://www.onlinepianist.com/virtual-piano to play some notes and see what pitch is displayed in the UI. 440Hz is A1, which is tone 33. See https://de.wikipedia.org/wiki/Kammerton
The link you mentioned uses a different standard.

Comment (achimmihca) [reply UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs]: For the keys named Y and C on the online-piano that you linked above, the UI shows me 57 as A3 and 69 as A4.

Comment (basisbit) [reply UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs]: Click on "Letter Notes" to have the virtual piano show key letters

Comment (achimmihca) [reply UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs]: Yes, the UI shows the correct reltive pitch (A for an A). Only question is if the octave is correct.

When I play a sine wave in Audacity with a perfect frequence of 440 Hz, then this is shown as A4 in the UI. This corresponds to the [Scientific Pitch Notation](https://en.wikipedia.org/wiki/Scientific_pitch_notation) (SPN). This is what I intended.

It is indeed confusing that there are different systems for the notation, I did not know this. Maybe I should add a comment to the code to make it clear. But it is not relevant for the game play anyway.

Comment (paradigmn) [reply UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs]: I am a bit confused at this point. Is the "+3" offset necessary or not?

Comment (achimmihca) [reply UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs]: BTW: The absolute name for a note is calculated in MidiUtils.cs

**Pull request merged**